devito/function.py

```diff
@@ -231,40 +232,44 @@ def _initialize_derivatives(self):
                     property(dx, 'Return the symbolic expression for '
                              'the centered first derivative wrt. '
                              'the %s dimension' % dim.name))
-
+            derivatives += ('d%s' % dim.name, )
             # First derivative, left
             dxl = partial(first_derivative, order=self.space_order,
```

Comment:
I'm concerned these checks still don't capture all possible derivatives that **cannot** be taken on the given ``Function``.

For **example**, think about the way we're constructing ``dxl`` and ``dxr``: we use the ``space_order`` of the function. The ``space_order`` is also used to build up the ``halo`` region. So here what's happening (to sum up) is that the number of left/right points required for ``dxl/dxr`` increases faster than the number of available points on the left/right  (see how ``_halo`` is computed in the ``Function`` constructor). 

So we should probably use ``_halo``, not ``space_order``. However, this checks would complicate this method, perhaps too much. Alternatively, we could change how ``first_derivative`` and ``generic_derivative`` behave: we could pass them a different parameter -- let's call it ``S`` -- representing "the number of grid points available on the left and on the right of a generic point x0" (basically the ``_halo``); then these function would return **only** the derivatives that can be taken given ``S``.

Does it make sense?

Reply:
I keep thinking about this because it causes a bug in the domain-allocation switch branch. So perhaps what's wrong is the way we're constructing ``dxl/dxr``. We should be using ``halo``, not ``space_order``, to generate "the possible maximum order left/right derivative given the available neighboring points".

Reply:
one last update:
if you cherry-pick this commit: https://github.com/opesci/devito/commit/77b900feb1f21b91ed122dd8068798deab02e6e5
you can get the number of left/right points available along a ``Function`` dimension via ``self._extent_halo[x]`` (where ``x`` is a ``Dimension`` and of course it can be any function indices). To be more precise, if you do ``self._extent_halo[x].left`` or ``self._extent_halo[x].right`` you actually get the left/right values.

Reply:
I have to disagree. The FD is defined by its `space_order`. I understand that the compiler would prefer it not it too, but we cannot start to generate stencils based on compiler or domain propeties.
Fd part shouldn't know anything about domain or halo ever, it generates the symbolci  fd scheme the user requires from the spatial order. If we start hacking into FD and change it in the middle of nowhere, this is gonna be problematic.

Devito is
- First generate a stencil
- Second generate code and run

If we start choosing what we run before we generate the stencil this will be a nightmare.